implementations/c/ockam/channel/channel_impl.c

```diff
@@ -0,0 +1,297 @@
+  p_ch->channel_reader->read = channel_read;
+  p_ch->channel_reader->ctx  = p_ch;
+
+  error = ockam_memory_alloc_zeroed(gp_ockam_channel_memory, (void**) &p_ch->channel_writer, sizeof(ockam_writer_t));
+  if (error) goto exit;
+  p_ch->channel_writer->write = channel_write;
+  p_ch->channel_writer->ctx   = p_ch;
+
+  p_ch->transport_reader = p_attrs->reader;
+  p_ch->transport_writer = p_attrs->writer;
+
+  error = ockam_xx_key_initialize(
+    &p_ch->key, gp_ockam_channel_memory, p_ch->vault, p_ch->channel_reader, p_ch->channel_writer);
+
+  p_ch->state = CHANNEL_STATE_M1;
+
+exit:
+  if (error) {
+    log_error(error, __func__);
+    if (p_ch) {
+      if (p_ch->channel_reader)
+        ockam_memory_free(gp_ockam_channel_memory, (void*) p_ch->channel_reader, sizeof(ockam_reader_t));
+      if (p_ch->channel_writer)
+        ockam_memory_free(gp_ockam_channel_memory, (void*) p_ch->channel_writer, sizeof(ockam_writer_t));
+    }
+  }
+  return 0;
+}
+
+ockam_error_t ockam_channel_connect(ockam_channel_t* p_ch, ockam_reader_t** p_reader, ockam_writer_t** p_writer)
+{
+  ockam_error_t error = 0;
+
+  error = ockam_key_initiate(&p_ch->key);
+  if (error) goto exit;
+  *p_reader = p_ch->channel_reader;
+  *p_writer = p_ch->channel_writer;
+
+exit:
+  if (error) log_error(error, __func__);
+  return error;
+}
+
+ockam_error_t ockam_channel_accept(ockam_channel_t* p_ch, ockam_reader_t** p_reader, ockam_writer_t** p_writer)
+{
+  ockam_error_t error = 0;
+  error               = ockam_key_respond(&p_ch->key);
+  if (error) goto exit;
+  *p_reader = p_ch->channel_reader;
+  *p_writer = p_ch->channel_writer;
+
+exit:
+  if (error) log_error(error, __func__);
+  return error;
+}
+
+ockam_error_t channel_read(void* ctx, uint8_t* p_clear_text, size_t clear_text_size, size_t* p_clear_text_length)
+{
+  ockam_error_t    error               = 0;
+  size_t           cipher_text_length  = 0;
+  size_t           encoded_text_length = 0;
+  uint8_t*         p_encoded           = g_encoded_text;
+  ockam_channel_t* p_ch                = (ockam_channel_t*) ctx;
+
+  error = ockam_read(p_ch->transport_reader, g_cipher_text, sizeof(g_cipher_text), &cipher_text_length);
+  if (error) goto exit;
+
+  error = channel_decrypt(
+    p_ch, g_cipher_text, cipher_text_length, g_encoded_text, sizeof(g_encoded_text), &encoded_text_length);
+  if (error) goto exit;
+
+  p_encoded = channel_deocde_header(p_ch, p_encoded);
+  if (NULL == p_encoded) {
+    error = CHANNEL_ERROR_NOT_IMPLEMENTED;
+    goto exit;
+  }
+
+  if (CHANNEL_STATE_SECURE == p_ch->state) {
+    error = channel_process_message(p_encoded, encoded_text_length, p_clear_text, p_clear_text_length);
+    if (error) goto exit;
+  } else {
+    codec_message_type_t message_type = *p_encoded++;
+    *p_clear_text_length              = encoded_text_length - (p_encoded - g_encoded_text);
+    ockam_memory_copy(gp_ockam_channel_memory, p_clear_text, p_encoded, *p_clear_text_length);
+    switch (p_ch->state) {
```

Comment:
It looks like this code relies on message order guarantee, but AFAIK underlying transport doesn't fulfil that requirement.